
<!DOCTYPE html>

<html class="writer-html5" data-content_root="./" lang="en">
<head>
<meta charset="utf-8"/><meta content="width=device-width, initial-scale=1" name="viewport"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<title>PLNmodels: Poisson lognormal models ‚Äî pyPLNmodels</title>
<link href="_static/pygments.css?v=b86133f3" rel="stylesheet" type="text/css"/>
<link href="_static/css/theme.css?v=e59714d7" rel="stylesheet" type="text/css"/>
<link href="_static/copybutton.css?v=76b2166b" rel="stylesheet" type="text/css"/>
<link href="_static/sphinx-codeautolink.css?v=125d5c1c" rel="stylesheet" type="text/css"/>
<link href="_static/autoclasstoc.css?v=08871587" rel="stylesheet" type="text/css"/>
<script src="_static/jquery.js?v=5d32c60e"></script>
<script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
<script src="_static/documentation_options.js?v=67224d9c"></script>
<script src="_static/doctools.js?v=9a2dae69"></script>
<script src="_static/sphinx_highlight.js?v=dc90522c"></script>
<script src="_static/clipboard.min.js?v=a7894cd8"></script>
<script src="_static/copybutton.js?v=f281be69"></script>
<script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script src="_static/js/theme.js"></script>
<link href="genindex.html" rel="index" title="Index"/>
<link href="search.html" rel="search" title="Search"/>
<link href="plndiag.html" rel="prev" title="PlnDiag"/>
</head>
<body class="wy-body-for-nav">
<div class="wy-grid-for-nav">
<nav class="wy-nav-side" data-toggle="wy-nav-shift">
<div class="wy-side-scroll">
<div class="wy-side-nav-search">
<a class="icon icon-home" href="index.html">
            pyPLNmodels
          </a>
<div role="search">
<form action="search.html" class="wy-form" id="rtd-search-form" method="get">
<input aria-label="Search docs" name="q" placeholder="Search docs" type="text"/>
<input name="check_keywords" type="hidden" value="yes"/>
<input name="area" type="hidden" value="default"/>
</form>
</div>
</div><div aria-label="Navigation menu" class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation">
<ul>
<li class="toctree-l1"><a class="reference internal" href="vignettes/index.html">My new Vignette</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Models</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pln.html">Pln</a></li>
<li class="toctree-l1"><a class="reference internal" href="zipln.html">ZIPln</a></li>
<li class="toctree-l1"><a class="reference internal" href="plnpca.html">PlnPCA</a></li>
<li class="toctree-l1"><a class="reference internal" href="plnpcacollection.html">PlnPCACollection</a></li>
<li class="toctree-l1"><a class="reference internal" href="plnar.html">PlnAR</a></li>
<li class="toctree-l1"><a class="reference internal" href="plnlda.html">PlnLDA</a></li>
<li class="toctree-l1"><a class="reference internal" href="plnnetwork.html">PlnNetwork</a></li>
<li class="toctree-l1"><a class="reference internal" href="plnnetworkcollection.html">PlnNetworkCollection</a></li>
<li class="toctree-l1"><a class="reference internal" href="plnmixture.html">PlnMixture</a></li>
<li class="toctree-l1"><a class="reference internal" href="plnmixturecollection.html">PlnMixtureCollection</a></li>
<li class="toctree-l1"><a class="reference internal" href="ziplnpca.html">ZIPlnPCA</a></li>
<li class="toctree-l1"><a class="reference internal" href="ziplnpcacollection.html">ZIPlnPCACollection</a></li>
<li class="toctree-l1"><a class="reference internal" href="plndiag.html">PlnDiag</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Overview</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">PLNmodels: Poisson lognormal models</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#getting-started">Getting started</a></li>
<li class="toctree-l2"><a class="reference internal" href="#installation">üõ† Installation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#package-installation">Package installation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#statistical-description">Statistical description</a></li>
<li class="toctree-l2"><a class="reference internal" href="#quickstart">‚ö°Ô∏è Quickstart</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#how-to-specify-a-model">How to specify a model</a></li>
<li class="toctree-l3"><a class="reference internal" href="#motivation">Motivation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#covariance-analysis-with-the-poisson-lognormal-model-aka-pln">Covariance analysis with the Poisson lognormal model (aka <code class="docutils literal notranslate"><span class="pre">Pln</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#dimension-reduction-with-the-pln-principal-component-analysis-aka-plnpca-and-plnpcacollection">Dimension reduction with the PLN Principal Component Analysis (aka <code class="docutils literal notranslate"><span class="pre">PlnPCA</span></code> and <code class="docutils literal notranslate"><span class="pre">PlnPCACollection</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#zero-inflation-with-the-zero-inflated-pln-model-aka-zipln-and-ziplnpca">Zero inflation with the Zero-Inflated PLN Model (aka <code class="docutils literal notranslate"><span class="pre">ZIPln</span></code> and <code class="docutils literal notranslate"><span class="pre">ZIPlnPCA</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#network-inference-with-the-plnnetwork-model">Network inference with the <code class="docutils literal notranslate"><span class="pre">PlnNetwork</span></code> model</a></li>
<li class="toctree-l3"><a class="reference internal" href="#supervised-clustering-with-the-plnlda-model">Supervised clustering with the <code class="docutils literal notranslate"><span class="pre">PlnLDA</span></code> model</a></li>
<li class="toctree-l3"><a class="reference internal" href="#unsupervised-clustering-with-the-plnmixture-model">Unsupervised clustering with the <code class="docutils literal notranslate"><span class="pre">PlnMixture</span></code> model</a></li>
<li class="toctree-l3"><a class="reference internal" href="#autoregressive-models-with-the-plnar-model">Autoregressive models with the <code class="docutils literal notranslate"><span class="pre">PlnAR</span></code> model</a></li>
<li class="toctree-l3"><a class="reference internal" href="#visualization">Visualization</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#contributing">üëê Contributing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#citations">‚ö°Ô∏è Citations</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Links</span></p>
<ul>
<li class="toctree-l1"><a class="reference external" href="https://pypi.org/project/pyPLNmodels/">‚Ü™ PyPI</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/PLN-team/pyPLNmodels">‚Ü™ GitHub</a></li>
</ul>
</div>
</div>
</nav>
<section class="wy-nav-content-wrap" data-toggle="wy-nav-shift"><nav aria-label="Mobile navigation menu" class="wy-nav-top">
<i class="fa fa-bars" data-toggle="wy-nav-top"></i>
<a href="index.html">pyPLNmodels</a>
</nav>
<div class="wy-nav-content">
<div class="rst-content">
<div aria-label="Page navigation" role="navigation">
<ul class="wy-breadcrumbs">
<li><a aria-label="Home" class="icon icon-home" href="index.html"></a></li>
<li class="breadcrumb-item active">PLNmodels: Poisson lognormal models</li>
<li class="wy-breadcrumbs-aside">
<a href="_sources/readme.rst.txt" rel="nofollow"> View page source</a>
</li>
</ul>
<hr/>
</div>
<div class="document" itemscope="itemscope" itemtype="http://schema.org/Article" role="main">
<div itemprop="articleBody">
<p><img alt="PyPI" src="https://img.shields.io/pypi/v/pyPLNmodels"/> <img alt="GitHub" src="https://img.shields.io/github/license/PLN-team/pyPLNmodels"/> <img alt="Python Versions" src="https://img.shields.io/badge/python-3.8%20%7C%203.9%20%7C%203.10%20%7C%203.11%20%7C%203.12%20%7C%203.13-blue"/> <img alt="Code Style: Black" src="https://img.shields.io/badge/code%20style-black-000000.svg"/> <img alt="Pylint" src="https://img.shields.io/badge/pylint-checked-brightgreen"/> <img alt="GPU Support" src="https://img.shields.io/badge/GPU-Supported-brightgreen"/> <a class="reference external" href="https://codecov.io/gh/PLN-team/pyPLNmodels"><img alt="codecov" src="https://codecov.io/gh/PLN-team/pyPLNmodels/graph/badge.svg?token=TNFROMLF9Z"/></a> <a class="reference external" href="https://github.com/PLN-team/pyPLNmodels"><img alt="last commit" src="https://img.shields.io/github/last-commit/PLN-team/pyPLNmodels"/></a></p>
<section id="plnmodels-poisson-lognormal-models">
<h1>PLNmodels: Poisson lognormal models<a class="headerlink" href="#plnmodels-poisson-lognormal-models" title="Link to this heading">ÔÉÅ</a></h1>
<blockquote>
<div><p>The Poisson lognormal model and its variants are used for the
analysis of multivariate count data. This package implements
efficient algorithms to extract meaningful insights from complex and
difficult-to-interpret multivariate count data. It is designed to
scale on large datasets, although it has memory limitations.
Possible fields of application include: - Genomics/transcriptomics
(e.g., the number of times a gene is expressed in a cell) - Ecology
(e.g., species abundances) - Metagenomics (e.g., the number of
Amplicon Sequence Variants detected in a sample) - Genetics (e.g.,
the number of crossovers (DNA exchanges) along a chromosome)</p>
<p>One of the main functionalities of this package is to normalize
count data to obtain more valuable insights. It also analyzes the
significance of each variable, their correlations, and the weight of
covariates (if available).</p>
</div></blockquote>
<p>The package documentation can be found
<a class="reference external" href="https://pln-team.github.io/pyPLNmodels/">here</a>.</p>
<section id="getting-started">
<h2>Getting started<a class="headerlink" href="#getting-started" title="Link to this heading">ÔÉÅ</a></h2>
<p><a class="reference external" href="https://github.com/PLN-team/pyPLNmodels/blob/main/Getting_started.ipynb">A notebook to get started can be found
here</a>.
If you need just a quick view of the package, see the quickstart next.
Note that an <code class="docutils literal notranslate"><span class="pre">R</span></code> version of the package is available
<a class="reference external" href="https://pln-team.github.io/PLNmodels/">here</a>.</p>
</section>
<section id="installation">
<h2>üõ† Installation<a class="headerlink" href="#installation" title="Link to this heading">ÔÉÅ</a></h2>
<p><strong>pyPLNmodels</strong> is available on
<a class="reference external" href="https://pypi.org/project/pyPLNmodels/">pypi</a>. The development
version is available on
<a class="reference external" href="https://github.com/PLN-team/pyPLNmodels">GitHub</a> and
<a class="reference external" href="https://gitlab.com/Bastien-mva/pyplnmodels">GitLab</a>.</p>
<section id="package-installation">
<h3>Package installation<a class="headerlink" href="#package-installation" title="Link to this heading">ÔÉÅ</a></h3>
<div class="highlight-sh notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>pyPLNmodels
</pre></div>
</div>
</section>
</section>
<section id="statistical-description">
<h2>Statistical description<a class="headerlink" href="#statistical-description" title="Link to this heading">ÔÉÅ</a></h2>
<p>For those unfamiliar with Poisson or Gaussian random variables, it‚Äôs not
necessary to delve into these statistical concepts. The key takeaway is
that this package analyzes multi-dimensional count data, extracting
significant information such as the mean, relationships with covariates,
and correlations between count variables, in a manner appropriate for
count data.</p>
<p>Consider <span class="math">\mathbf Y</span> a count matrix (denoted as <code class="docutils literal notranslate"><span class="pre">endog</span></code> in the
package) consisting of <span class="math">n</span> rows and <span class="math">p</span> columns. It is
assumed that each individual <span class="math">\mathbf Y_i</span>, that is the
<span class="math">i^{\text{th}}</span> row of <span class="math">\mathbf Y</span>, is independent from the
others and follows a Poisson lognormal distribution:</p>
<div class="math notranslate nohighlight">
\[\mathbf Y_{i}\sim \mathcal P(\exp(\mathbf Z_{i})), \quad \mathbf Z_i \sim
\mathcal N(\mathbf o_i + \mathbf B ^{\top} \mathbf x_i, \mathbf \Sigma), (\text{PLN-equation})\]</div>
<p>where <span class="math">\mathbf x_i \in \mathbb R^d</span> (<code class="docutils literal notranslate"><span class="pre">exog</span></code>) and
<span class="math">\mathbf o_i \in \mathbb R^p</span> (<code class="docutils literal notranslate"><span class="pre">offsets</span></code>) are user-specified
covariates and offsets. The matrix <span class="math">\mathbf B</span> is a
<span class="math">d\times p</span> matrix of regression coefficients and
<span class="math">\mathbf \Sigma</span> is a <span class="math">p\times p</span> covariance matrix. The
goal is to estimate the parameters <span class="math">\mathbf B</span> and
<span class="math">\mathbf \Sigma</span>, denoted as <code class="docutils literal notranslate"><span class="pre">coef</span></code> and <code class="docutils literal notranslate"><span class="pre">covariance</span></code> in the
package, respectively.</p>
<p>The PLN model described in the PLN-equation is the building block of
many different statistical tasks adequate for count data, by modifying
the <span class="math">Z_i</span> latent variables. The package implements:</p>
<ul class="simple">
<li><p>Covariance analysis (<code class="docutils literal notranslate"><span class="pre">Pln</span></code>)</p></li>
<li><p>Dimension reduction (<code class="docutils literal notranslate"><span class="pre">PlnPCA</span></code> and <code class="docutils literal notranslate"><span class="pre">PlnPCACollection</span></code>)</p></li>
<li><p>Zero-inflation (<code class="docutils literal notranslate"><span class="pre">ZIPln</span></code>)</p></li>
<li><p>Autoregressive models (<code class="docutils literal notranslate"><span class="pre">PlnAR</span></code>)</p></li>
<li><p>Supervised clustering (<code class="docutils literal notranslate"><span class="pre">PlnLDA</span></code>)</p></li>
<li><p>Unsupervised clustering (<code class="docutils literal notranslate"><span class="pre">PlnMixture</span></code>)</p></li>
<li><p>Network inference (<code class="docutils literal notranslate"><span class="pre">PlnNetwork</span></code>)</p></li>
<li><p>Zero-inflation and dimension reduction (<code class="docutils literal notranslate"><span class="pre">ZIPlnPCA</span></code>)</p></li>
<li><p>Variance estimation (<code class="docutils literal notranslate"><span class="pre">PlnDiag</span></code>)</p></li>
</ul>
<p>A normalization procedure adequate to count data can be applied by
extracting the <code class="docutils literal notranslate"><span class="pre">latent_variables</span></code> <span class="math">\mathbf Z_i</span> once the
parameters are learned.</p>
</section>
<section id="quickstart">
<h2>‚ö°Ô∏è Quickstart<a class="headerlink" href="#quickstart" title="Link to this heading">ÔÉÅ</a></h2>
<p>The package comes with a single-cell RNA sequencing dataset to present
the functionalities:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pyPLNmodels</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_scrna</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">load_scrna</span><span class="p">()</span>
</pre></div>
</div>
<p>This dataset contains the number of occurrences of each gene in each
cell in <code class="docutils literal notranslate"><span class="pre">data["endog"]</span></code>. Each cell is labelled by its cell-type in
<code class="docutils literal notranslate"><span class="pre">data["labels"]</span></code>.</p>
<section id="how-to-specify-a-model">
<h3>How to specify a model<a class="headerlink" href="#how-to-specify-a-model" title="Link to this heading">ÔÉÅ</a></h3>
<p>Each model can be specified in two distinct manners:</p>
<ul>
<li><p>by formula (similar to R), where a data frame is passed and the
formula is specified using the <code class="docutils literal notranslate"><span class="pre">from_formula</span></code> initialization:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pyPLNmodels</span><span class="w"> </span><span class="kn">import</span> <a class="sphinx-codeautolink-a" href="pln.html#pyPLNmodels.Pln" title="pyPLNmodels.models.pln.Pln"><span class="n">Pln</span></a>
<span class="n">pln</span> <span class="o">=</span> <span class="n">Pln</span><span class="o">.</span><span class="n">from_formula</span><span class="p">(</span><span class="s2">"endog ~ 1  + labels "</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</li>
</ul>
<p>We rely on the <a class="reference external" href="https://github.com/pydata/patsy">patsy</a> package for
the formula parsing.</p>
<ul>
<li><p>by specifying the <code class="docutils literal notranslate"><span class="pre">endog</span></code>, <code class="docutils literal notranslate"><span class="pre">exog</span></code>, and <code class="docutils literal notranslate"><span class="pre">offsets</span></code> matrices
directly:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="n">endog</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">"endog"</span><span class="p">]</span>
<span class="n">exog</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">"labels"</span><span class="p">]</span>
<span class="n">offsets</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">endog</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
<span class="n">pln</span> <span class="o">=</span> <span class="n">Pln</span><span class="p">(</span><span class="n">endog</span><span class="o">=</span><span class="n">endog</span><span class="p">,</span> <span class="n">exog</span><span class="o">=</span><span class="n">exog</span><span class="p">,</span> <span class="n">offsets</span><span class="o">=</span><span class="n">offsets</span><span class="p">)</span>
</pre></div>
</div>
</li>
</ul>
<p>The parameters <code class="docutils literal notranslate"><span class="pre">exog</span></code> and <code class="docutils literal notranslate"><span class="pre">offsets</span></code> are optional. By default,
<code class="docutils literal notranslate"><span class="pre">exog</span></code> is set to represent an intercept, which is a vector of ones.
Similarly, <code class="docutils literal notranslate"><span class="pre">offsets</span></code> defaults to a matrix of zeros. The <code class="docutils literal notranslate"><span class="pre">offsets</span></code>
should be on the scale of the log of the counts.</p>
</section>
<section id="motivation">
<h3>Motivation<a class="headerlink" href="#motivation" title="Link to this heading">ÔÉÅ</a></h3>
<p>The count data is often very noisy, and inferring the latent variables
<span class="math">Z_i</span> may reduce noise and increase signal. Suppose we try to
infer the cell type of each cell, using Linear Discriminant Analysis
(LDA):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.model_selection</span><span class="w"> </span><span class="kn">import</span> <span class="n">train_test_split</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.discriminant_analysis</span><span class="w"> </span><span class="kn">import</span> <span class="n">LinearDiscriminantAnalysis</span> <span class="k">as</span> <span class="n">LDA</span>

<span class="k">def</span><span class="w"> </span><span class="nf">get_classif_error</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">y</span><span class="p">):</span>
    <span class="n">data_train</span><span class="p">,</span> <span class="n">data_test</span><span class="p">,</span> <span class="n">y_train</span><span class="p">,</span> <span class="n">y_test</span> <span class="o">=</span> <span class="n">train_test_split</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">test_size</span><span class="o">=</span><span class="mf">0.33</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
    <span class="n">lda</span> <span class="o">=</span> <span class="n">LDA</span><span class="p">()</span>
    <span class="n">lda</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">data_train</span><span class="p">,</span> <span class="n">y_train</span><span class="p">)</span>
    <span class="n">y_pred</span> <span class="o">=</span> <span class="n">lda</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">data_test</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">y_pred</span> <span class="o">!=</span> <span class="n">y_test</span><span class="p">)</span>
</pre></div>
</div>
<p>Here is the classification error of the raw counts:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">load_scrna</span><span class="p">(</span><span class="n">n_samples</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="n">get_classif_error</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">"endog"</span><span class="p">],</span> <span class="n">data</span><span class="p">[</span><span class="s2">"labels"</span><span class="p">])</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mf">0.31</span>
</pre></div>
</div>
<p>And here is the classification error of the latent variables
<span class="math">Z_i</span>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">get_classif_error</span><span class="p">(</span><span class="n">Pln</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">"endog"</span><span class="p">])</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span><span class="o">.</span><span class="n">latent_variables</span><span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="s2">"labels"</span><span class="p">])</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mf">0.17</span>
</pre></div>
</div>
</section>
<section id="covariance-analysis-with-the-poisson-lognormal-model-aka-pln">
<h3>Covariance analysis with the Poisson lognormal model (aka <code class="docutils literal notranslate"><span class="pre">Pln</span></code>)<a class="headerlink" href="#covariance-analysis-with-the-poisson-lognormal-model-aka-pln" title="Link to this heading">ÔÉÅ</a></h3>
<p>This is the building-block of the models implemented in this package. It
fits a Poisson lognormal model to the data:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pln</span> <span class="o">=</span> <span class="n">Pln</span><span class="o">.</span><span class="n">from_formula</span><span class="p">(</span><span class="s2">"endog ~ 1  + labels "</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">)</span>
<span class="n">pln</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pln</span><span class="p">)</span>
<span class="n">transformed_data</span> <span class="o">=</span> <span class="n">pln</span><span class="o">.</span><span class="n">transform</span><span class="p">()</span>
<span class="n">pln</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="dimension-reduction-with-the-pln-principal-component-analysis-aka-plnpca-and-plnpcacollection">
<h3>Dimension reduction with the PLN Principal Component Analysis (aka <code class="docutils literal notranslate"><span class="pre">PlnPCA</span></code> and <code class="docutils literal notranslate"><span class="pre">PlnPCACollection</span></code>)<a class="headerlink" href="#dimension-reduction-with-the-pln-principal-component-analysis-aka-plnpca-and-plnpcacollection" title="Link to this heading">ÔÉÅ</a></h3>
<p>This model excels in dimension reduction and is capable of scaling to
high-dimensional count data (<span class="math">p &gt;&gt; 1</span>), by constraining the
covariance matrix <span class="math">\Sigma</span> to be of low rank (the larger the rank,
the slower the model but the better the approximation). The user may
specify the rank when creating the <code class="docutils literal notranslate"><span class="pre">PlnPCA</span></code> object:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pyPLNmodels</span><span class="w"> </span><span class="kn">import</span> <a class="sphinx-codeautolink-a" href="plnpca.html#pyPLNmodels.PlnPCA" title="pyPLNmodels.models.plnpca.PlnPCA"><span class="n">PlnPCA</span></a>
<span class="n">pca</span> <span class="o">=</span> <span class="n">PlnPCA</span><span class="o">.</span><span class="n">from_formula</span><span class="p">(</span><span class="s2">"endog ~ 1  + labels "</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">rank</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
</pre></div>
</div>
<p>Multiple ranks can be simultaneously tested within a single object
(<code class="docutils literal notranslate"><span class="pre">PlnPCACollection</span></code>), and select the optimal model.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pyPLNmodels</span><span class="w"> </span><span class="kn">import</span> <a class="sphinx-codeautolink-a" href="plnpcacollection.html#pyPLNmodels.PlnPCACollection" title="pyPLNmodels.models.plnpcacollection.PlnPCACollection"><span class="n">PlnPCACollection</span></a>
<span class="n">pca_col</span> <span class="o">=</span> <span class="n">PlnPCACollection</span><span class="o">.</span><span class="n">from_formula</span><span class="p">(</span><span class="s2">"endog ~ 1  + labels "</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">ranks</span> <span class="o">=</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">])</span>
<span class="n">pca_col</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pca_col</span><span class="p">)</span>
<span class="n">pca_col</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">best_pca</span> <span class="o">=</span> <span class="n">pca_col</span><span class="o">.</span><span class="n">best_model</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">best_pca</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="zero-inflation-with-the-zero-inflated-pln-model-aka-zipln-and-ziplnpca">
<h3>Zero inflation with the Zero-Inflated PLN Model (aka <code class="docutils literal notranslate"><span class="pre">ZIPln</span></code> and <code class="docutils literal notranslate"><span class="pre">ZIPlnPCA</span></code>)<a class="headerlink" href="#zero-inflation-with-the-zero-inflated-pln-model-aka-zipln-and-ziplnpca" title="Link to this heading">ÔÉÅ</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">ZIPln</span></code> model, a variant of the PLN model, is designed to handle
zero inflation in the data. It is defined as follows:</p>
<div class="math notranslate nohighlight">
\[Y_{ij}\sim \mathcal W_{ij} \times  P(\exp(Z_{ij})), \quad \mathbf Z_i \sim \mathcal N(\mathbf o_i + \mathbf B ^{\top} \mathbf x_i, \mathbf \Sigma), \quad W_{ij} \sim \mathcal B(\sigma( \mathbf x_i^{0^{\top}}\mathbf B^0_j))\]</div>
<p>This model is particularly beneficial when the data contains a
significant number of zeros. It incorporates additional covariates for
the zero inflation coefficient, which are specified following the pipe
<code class="docutils literal notranslate"><span class="pre">|</span></code> symbol in the formula or via the <code class="docutils literal notranslate"><span class="pre">exog_inflation</span></code> keyword. If
not specified, it is set to the covariates for the Poisson part.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pyPLNmodels</span><span class="w"> </span><span class="kn">import</span> <a class="sphinx-codeautolink-a" href="zipln.html#pyPLNmodels.ZIPln" title="pyPLNmodels.models.zipln.ZIPln"><span class="n">ZIPln</span></a>
<span class="n">zi</span> <span class="o">=</span> <span class="n">ZIPln</span><span class="o">.</span><span class="n">from_formula</span><span class="p">(</span><span class="s2">"endog ~ 1 | 1 + labels"</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">zi</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">"Transformed data shape: "</span><span class="p">,</span> <span class="n">zi</span><span class="o">.</span><span class="n">transform</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">z_latent_variables</span> <span class="o">=</span> <span class="n">zi</span><span class="o">.</span><span class="n">transform</span><span class="p">()</span>
<span class="n">w_latent_variables</span> <span class="o">=</span> <span class="n">zi</span><span class="o">.</span><span class="n">latent_prob</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">r</span><span class="s1">'$Z$ latent variables shape'</span><span class="p">,</span> <span class="n">z_latent_variables</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">r</span><span class="s1">'$W$ latent variables shape'</span><span class="p">,</span> <span class="n">w_latent_variables</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
<p>Similar to the <code class="docutils literal notranslate"><span class="pre">PlnPCA</span></code> model, the <code class="docutils literal notranslate"><span class="pre">ZIPlnPCA</span></code> model is capable of
dimension reduction.</p>
</section>
<section id="network-inference-with-the-plnnetwork-model">
<h3>Network inference with the <code class="docutils literal notranslate"><span class="pre">PlnNetwork</span></code> model<a class="headerlink" href="#network-inference-with-the-plnnetwork-model" title="Link to this heading">ÔÉÅ</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">PlnNetwork</span></code> model is designed to infer the network structure of
the data. It creates a network where the nodes are the count variables
and the edges represent the correlation between them. The sparsity of
the network is ensured via the <code class="docutils literal notranslate"><span class="pre">penalty</span></code> keyword. The larger the
penalty, the sparser the network.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pyPLNmodels</span><span class="w"> </span><span class="kn">import</span> <a class="sphinx-codeautolink-a" href="plnnetwork.html#pyPLNmodels.PlnNetwork" title="pyPLNmodels.models.plnnetwork.PlnNetwork"><span class="n">PlnNetwork</span></a>
<span class="n">net</span> <span class="o">=</span> <span class="n">PlnNetwork</span><span class="o">.</span><span class="n">from_formula</span><span class="p">(</span><span class="s2">"endog ~ 1  + labels "</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">penalty</span> <span class="o">=</span> <span class="mi">200</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
<span class="n">net</span><span class="o">.</span><span class="n">viz_network</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">net</span><span class="o">.</span><span class="n">network</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="supervised-clustering-with-the-plnlda-model">
<h3>Supervised clustering with the <code class="docutils literal notranslate"><span class="pre">PlnLDA</span></code> model<a class="headerlink" href="#supervised-clustering-with-the-plnlda-model" title="Link to this heading">ÔÉÅ</a></h3>
<p>One can do supervised clustering using Linear Discriminant Analysis
designed for count data.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pyPLNmodels</span><span class="w"> </span><span class="kn">import</span> <a class="sphinx-codeautolink-a" href="plnlda.html#pyPLNmodels.PlnLDA" title="pyPLNmodels.models.plnlda.PlnLDA"><span class="n">PlnLDA</span></a><span class="p">,</span> <span class="n">plot_confusion_matrix</span>
<span class="n">endog_train</span><span class="p">,</span> <span class="n">endog_test</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">"endog"</span><span class="p">][:</span><span class="mi">500</span><span class="p">],</span> <span class="n">data</span><span class="p">[</span><span class="s2">"endog"</span><span class="p">][</span><span class="mi">500</span><span class="p">:]</span>
<span class="n">labels_train</span><span class="p">,</span> <span class="n">labels_test</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">"labels"</span><span class="p">][:</span><span class="mi">500</span><span class="p">],</span> <span class="n">data</span><span class="p">[</span><span class="s2">"labels"</span><span class="p">][</span><span class="mi">500</span><span class="p">:]</span>
<span class="n">lda</span> <span class="o">=</span> <a class="sphinx-codeautolink-a" href="plnlda.html#pyPLNmodels.PlnLDA" title="pyPLNmodels.models.plnlda.PlnLDA"><span class="n">PlnLDA</span></a><span class="p">(</span><span class="n">endog_train</span><span class="p">,</span> <span class="n">clusters</span><span class="o">=</span><span class="n">labels_train</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
<span class="n">pred_test</span> <span class="o">=</span> <span class="n">lda</span><span class="o">.</span><span class="n">predict_clusters</span><span class="p">(</span><span class="n">endog_test</span><span class="p">)</span>
<span class="n">plot_confusion_matrix</span><span class="p">(</span><span class="n">pred_test</span><span class="p">,</span> <span class="n">labels_test</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="unsupervised-clustering-with-the-plnmixture-model">
<h3>Unsupervised clustering with the <code class="docutils literal notranslate"><span class="pre">PlnMixture</span></code> model<a class="headerlink" href="#unsupervised-clustering-with-the-plnmixture-model" title="Link to this heading">ÔÉÅ</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pyPLNmodels</span><span class="w"> </span><span class="kn">import</span> <a class="sphinx-codeautolink-a" href="plnmixture.html#pyPLNmodels.PlnMixture" title="pyPLNmodels.models.plnmixture.PlnMixture"><span class="n">PlnMixture</span></a>
<span class="n">mixture</span> <span class="o">=</span> <span class="n">PlnMixture</span><span class="o">.</span><span class="n">from_formula</span><span class="p">(</span><span class="s2">"endog ~ 0 "</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">n_cluster</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
<span class="n">mixture</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">clusters</span> <span class="o">=</span> <span class="n">mixture</span><span class="o">.</span><span class="n">clusters</span>
<span class="n">plot_confusion_matrix</span><span class="p">(</span><span class="n">clusters</span><span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="s2">"labels"</span><span class="p">])</span>
</pre></div>
</div>
</section>
<section id="autoregressive-models-with-the-plnar-model">
<h3>Autoregressive models with the <code class="docutils literal notranslate"><span class="pre">PlnAR</span></code> model<a class="headerlink" href="#autoregressive-models-with-the-plnar-model" title="Link to this heading">ÔÉÅ</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">PlnAR</span></code> model is designed to handle time series data. It is a
simple (one step) autoregressive model that can be used to predict the
next time point. (This assumes the endog variable is a time series,
which is not the case in the example below)</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pyPLNmodels</span><span class="w"> </span><span class="kn">import</span> <a class="sphinx-codeautolink-a" href="plnar.html#pyPLNmodels.PlnAR" title="pyPLNmodels.models.plnar.PlnAR"><span class="n">PlnAR</span></a>
<span class="n">ar</span> <span class="o">=</span> <span class="n">PlnAR</span><span class="o">.</span><span class="n">from_formula</span><span class="p">(</span><span class="s2">"endog ~ 1  + labels "</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
<span class="n">ar</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="visualization">
<h3>Visualization<a class="headerlink" href="#visualization" title="Link to this heading">ÔÉÅ</a></h3>
<p>The package is equipped with a set of visualization functions designed
to help the user interpret the data. The <code class="docutils literal notranslate"><span class="pre">viz</span></code> function conducts
<code class="docutils literal notranslate"><span class="pre">PCA</span></code> on the latent variables. The <code class="docutils literal notranslate"><span class="pre">remove_exog_effect</span></code> keyword
removes the covariates‚Äô effect specified in the model when set to
<code class="docutils literal notranslate"><span class="pre">True</span></code>. Much more functionalities, depending on the model, are
available. One can see the full list of available functions in the
documentation and by printing the model:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">pln</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pca</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pca_col</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">zi</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">net</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">lda</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mixture</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ar</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="contributing">
<h2>üëê Contributing<a class="headerlink" href="#contributing" title="Link to this heading">ÔÉÅ</a></h2>
<p>Feel free to contribute, but read the
<a class="reference external" href="https://github.com/PLN-team/pyPLNmodels/blob/main/CONTRIBUTING.md">CONTRIBUTING.md</a>
first. A public roadmap will be available soon.</p>
</section>
<section id="citations">
<h2>‚ö°Ô∏è Citations<a class="headerlink" href="#citations" title="Link to this heading">ÔÉÅ</a></h2>
<p>Please cite our work using the following references:</p>
<ul class="simple">
<li><p>B. Batardiere, J.Kwon, J.Chiquet: <em>pyPLNmodels: A Python package to
analyze multivariate high-dimensional count data.</em>
<a class="reference external" href="https://joss.theoj.org/papers/10.21105/joss.06969">pdf</a></p></li>
<li><p>J. Chiquet, M. Mariadassou and S. Robin: <em>Variational inference for
probabilistic Poisson PCA, the Annals of Applied Statistics, 12:
2674‚Äì2698, 2018.</em> <a class="reference external" href="http://dx.doi.org/10.1214/18%2DAOAS1177">pdf</a></p></li>
<li><p>B. Batardiere, J.Chiquet, M.Mariadassou: <em>Zero-inflation in the
Multivariate Poisson Lognormal Family.</em>
<a class="reference external" href="https://arxiv.org/abs/2405.14711">pdf</a></p></li>
<li><p>B. Batardiere, J.Chiquet, M.Mariadassou: <em>Evaluating Parameter
Uncertainty in the Poisson Lognormal Model with Corrected Variational
Estimators.</em> <a class="reference external" href="https://arxiv.org/abs/2411.08524">pdf</a></p></li>
<li><p>J. Chiquet, M. Mariadassou, S. Robin: <em>The Poisson-Lognormal Model as
a Versatile Framework for the Joint Analysis of Species Abundances.</em>
<a class="reference external" href="https://www.frontiersin.org/journals/ecology-and-evolution/articles/10.3389/fevo.2021.588292/full">pdf</a></p></li>
<li><p>J. Chiquet, S. Robin, M. Mariadassou: <em>Variational Inference for
sparse network reconstruction from count data</em>
<a class="reference external" href="https://proceedings.mlr.press/v97/chiquet19a.html">pdf</a></p></li>
</ul>
</section>
</section>
</div>
</div>
<footer><div aria-label="Footer" class="rst-footer-buttons" role="navigation">
<a accesskey="p" class="btn btn-neutral float-left" href="plndiag.html" rel="prev" title="PlnDiag"><span aria-hidden="true" class="fa fa-arrow-circle-left"></span> Previous</a>
</div>
<hr/>
<div role="contentinfo">
<p>¬© Copyright 2023, Bastien Batardi√®re, Julien Chiquet, Joon Kwon.</p>
</div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
</div>
</div>
</section>
</div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
</body>
</html>